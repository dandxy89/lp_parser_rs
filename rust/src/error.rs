//! Comprehensive error types for LP parsing operations.
//!
//! This module provides a structured error hierarchy with detailed context
//! information for debugging and error handling in LP file parsing.

use thiserror::Error;

/// Comprehensive error type for LP parsing operations.
///
/// This error type provides detailed context about parsing failures,
/// including location information and specific error conditions.
#[derive(Error, Debug, Clone, PartialEq, Eq)]
pub enum LpParseError {
    /// Invalid or malformed constraint syntax
    #[error("Invalid constraint syntax at position {position}: {context}")]
    ConstraintSyntax { position: usize, context: String },

    /// Invalid or malformed objective syntax
    #[error("Invalid objective syntax at position {position}: {context}")]
    ObjectiveSyntax { position: usize, context: String },

    /// Unknown or invalid variable type specification
    #[error("Unknown variable type '{var_type}' for variable '{variable}'")]
    UnknownVariableType { variable: String, var_type: String },

    /// Reference to an undefined variable
    #[error("Undefined variable '{variable}' referenced in {context}")]
    UndefinedVariable { variable: String, context: String },

    /// Duplicate definition of a component
    #[error("Duplicate {component_type} '{name}' defined")]
    DuplicateDefinition { component_type: String, name: String },

    /// Invalid numerical value or format
    #[error("Invalid number format '{value}' at position {position}")]
    InvalidNumber { value: String, position: usize },

    /// Missing required section in LP file
    #[error("Missing required section: {section}")]
    MissingSection { section: String },

    /// Invalid bound specification
    #[error("Invalid bounds for variable '{variable}': {details}")]
    InvalidBounds { variable: String, details: String },

    /// Invalid SOS constraint specification
    #[error("Invalid SOS constraint '{constraint}': {details}")]
    InvalidSosConstraint { constraint: String, details: String },

    /// Validation error for logical consistency
    #[error("Validation error: {message}")]
    ValidationError { message: String },

    /// Generic parsing error with context
    #[error("Parse error at position {position}: {message}")]
    ParseError { position: usize, message: String },

    /// File I/O related errors
    #[error("File I/O error: {message}")]
    IoError { message: String },

    /// Internal parser state errors
    #[error("Internal parser error: {message}")]
    InternalError { message: String },
}

impl LpParseError {
    /// Create a new constraint syntax error
    pub fn constraint_syntax(position: usize, context: impl Into<String>) -> Self {
        Self::ConstraintSyntax { position, context: context.into() }
    }

    /// Create a new objective syntax error
    pub fn objective_syntax(position: usize, context: impl Into<String>) -> Self {
        Self::ObjectiveSyntax { position, context: context.into() }
    }

    /// Create a new unknown variable type error
    pub fn unknown_variable_type(variable: impl Into<String>, var_type: impl Into<String>) -> Self {
        Self::UnknownVariableType { variable: variable.into(), var_type: var_type.into() }
    }

    /// Create a new undefined variable error
    pub fn undefined_variable(variable: impl Into<String>, context: impl Into<String>) -> Self {
        Self::UndefinedVariable { variable: variable.into(), context: context.into() }
    }

    /// Create a new duplicate definition error
    pub fn duplicate_definition(component_type: impl Into<String>, name: impl Into<String>) -> Self {
        Self::DuplicateDefinition { component_type: component_type.into(), name: name.into() }
    }

    /// Create a new invalid number error
    pub fn invalid_number(value: impl Into<String>, position: usize) -> Self {
        Self::InvalidNumber { value: value.into(), position }
    }

    /// Create a new missing section error
    pub fn missing_section(section: impl Into<String>) -> Self {
        Self::MissingSection { section: section.into() }
    }

    /// Create a new invalid bounds error
    pub fn invalid_bounds(variable: impl Into<String>, details: impl Into<String>) -> Self {
        Self::InvalidBounds { variable: variable.into(), details: details.into() }
    }

    /// Create a new invalid SOS constraint error
    pub fn invalid_sos_constraint(constraint: impl Into<String>, details: impl Into<String>) -> Self {
        Self::InvalidSosConstraint { constraint: constraint.into(), details: details.into() }
    }

    /// Create a new validation error
    pub fn validation_error(message: impl Into<String>) -> Self {
        Self::ValidationError { message: message.into() }
    }

    /// Create a new parse error
    pub fn parse_error(position: usize, message: impl Into<String>) -> Self {
        Self::ParseError { position, message: message.into() }
    }

    /// Create a new I/O error
    pub fn io_error(message: impl Into<String>) -> Self {
        Self::IoError { message: message.into() }
    }

    /// Create a new internal error
    pub fn internal_error(message: impl Into<String>) -> Self {
        Self::InternalError { message: message.into() }
    }
}

/// Convert from nom parsing errors to our custom error type
impl<'a> From<nom::Err<nom::error::Error<&'a str>>> for LpParseError {
    fn from(err: nom::Err<nom::error::Error<&'a str>>) -> Self {
        match err {
            nom::Err::Incomplete(_) => Self::parse_error(0, "Incomplete input"),
            nom::Err::Error(e) | nom::Err::Failure(e) => Self::parse_error(e.input.as_ptr() as usize, format!("Parse error: {:?}", e.code)),
        }
    }
}

/// Convert from standard I/O errors
impl From<std::io::Error> for LpParseError {
    fn from(err: std::io::Error) -> Self {
        Self::io_error(err.to_string())
    }
}

/// Convert from boxed errors (used by CSV module)
impl From<Box<dyn std::error::Error>> for LpParseError {
    fn from(err: Box<dyn std::error::Error>) -> Self {
        Self::io_error(err.to_string())
    }
}

/// Result type alias for LP parsing operations
pub type LpResult<T> = Result<T, LpParseError>;

/// Context extension trait for adding location information to errors
pub trait ErrorContext<T> {
    /// Add position context to an error
    fn with_position(self, position: usize) -> LpResult<T>;

    /// Add general context to an error
    fn with_context(self, context: &str) -> LpResult<T>;
}

impl<T> ErrorContext<T> for Result<T, LpParseError> {
    fn with_position(self, position: usize) -> Self {
        self.map_err(|mut err| {
            if let LpParseError::ParseError { position: ref mut pos, .. } = &mut err {
                *pos = position;
            }
            err
        })
    }

    fn with_context(self, context: &str) -> Self {
        self.map_err(|err| match err {
            LpParseError::ParseError { position, message } => LpParseError::parse_error(position, format!("{context}: {message}")),
            other => other,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_error_creation() {
        let err = LpParseError::constraint_syntax(42, "missing operator");
        assert_eq!(err.to_string(), "Invalid constraint syntax at position 42: missing operator");
    }

    #[test]
    fn test_error_context() {
        let result: LpResult<()> = Err(LpParseError::parse_error(10, "test error"));
        let with_context = result.with_context("parsing constraint");

        assert!(with_context.is_err());
        assert!(with_context.unwrap_err().to_string().contains("parsing constraint"));
    }

    #[test]
    fn test_io_error_conversion() {
        let io_err = std::io::Error::new(std::io::ErrorKind::NotFound, "file not found");
        let lp_err: LpParseError = io_err.into();

        match lp_err {
            LpParseError::IoError { message } => assert!(message.contains("file not found")),
            _ => panic!("Expected IoError"),
        }
    }
}
