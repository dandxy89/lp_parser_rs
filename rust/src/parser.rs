//! File I/O and content parsing utilities.
//!
//! This module provides functionality for reading LP files from disk
//! and preparing their contents for parsing.
//!

use std::fs::File;
use std::io::{BufReader, Read as _};
use std::path::Path;

use crate::error::{LpParseError, LpResult};

#[inline]
/// Parses the contents of a file at the given path into a string.
///
/// # Arguments
///
/// * `path` - A reference to a `Path` that represents the file path to be parsed.
///
/// # Returns
///
/// A `Result` containing the file contents as a `String` if successful, or an `LpParseError`
/// if the file cannot be opened or read.
///
/// # Errors
///
/// Returns an `LpParseError::IoError` if the file cannot be opened or read.
pub fn parse_file(path: &Path) -> LpResult<String> {
    let file = File::open(path).map_err(|e| LpParseError::io_error(format!("Failed to open file '{}': {}", path.display(), e)))?;

    let mut buf_reader = BufReader::new(file);
    let mut contents = String::new();

    buf_reader
        .read_to_string(&mut contents)
        .map_err(|e| LpParseError::io_error(format!("Failed to read file '{}': {}", path.display(), e)))?;

    Ok(contents)
}
