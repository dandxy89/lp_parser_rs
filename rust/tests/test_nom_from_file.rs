//! This Rust code snippet defines a function to read LP (Linear Programming) files from a resources
//! directory and parse them using the `lp_parser_rs` library.
//!
//! It includes several test functions and a macro to generate tests for various LP files, ensuring they can be parsed correctly.
//!

use std::error::Error;
use std::path::PathBuf;

use lp_parser_rs::parser::parse_file;
use lp_parser_rs::problem::LpProblem;

fn read_file_from_resources(file_name: &str) -> Result<String, Box<dyn Error + 'static>> {
    let mut file_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    file_path.push(format!("resources/{file_name}"));
    let contents = parse_file(&file_path)?;

    Ok(contents)
}

#[test]
fn invalid() {
    let input = read_file_from_resources("invalid.lp").expect("failed to read file from resources");
    assert!(LpProblem::parse(&input).is_err());
}

#[test]
fn nom_fit2d() {
    let input = read_file_from_resources("fit2d.lp").expect("failed to read file from resources");
    let parsed = LpProblem::parse(&input).expect("failed to parse LPProblem");
    assert_eq!(parsed.objective_count(), 1);
    assert_eq!(parsed.constraint_count(), 25);
    assert_eq!(parsed.variable_count(), 10500);

    #[cfg(feature = "serde")]
    insta::assert_yaml_snapshot!(parsed, {
        ".variables" => insta::sorted_redaction(),
        ".objectives" => insta::sorted_redaction(),
        ".constraints" => insta::sorted_redaction(),
    });
}

#[ignore]
#[test]
/// A test used my contributors for assisting with issues
fn nom_debug() {
    let input = read_file_from_resources("infile_comments.lp").expect("failed to read file from resources");
    assert!(LpProblem::parse(&input).is_ok());
}

#[macro_export]
macro_rules! generate_test {
    ($test_name:ident, $file:expr) => {
        #[test]
        #[allow(unused_variables)]
        fn $test_name() {
            let input = read_file_from_resources($file).unwrap();
            let parsed = LpProblem::parse(&input).expect("failed to parse LPProblem");

            #[cfg(feature = "serde")]
            insta::assert_yaml_snapshot!(parsed, {
                ".variables" => insta::sorted_redaction(),
                ".objectives" => insta::sorted_redaction(),
                ".constraints" => insta::sorted_redaction(),
            });
        }
    };
}

generate_test!(pulp, "pulp.lp");
generate_test!(pulp2, "pulp2.lp");
generate_test!(limbo, "limbo.lp");
// generate_test!(semi_continuous, "semi_continuous.lp");
generate_test!(sos, "sos.lp");
generate_test!(test, "test.lp");
generate_test!(test2, "test2.lp");
generate_test!(empty_bounds, "empty_bounds.lp");
generate_test!(blank_lines, "blank_lines.lp");
// generate_test!(optional_labels, "optional_labels.lp");
generate_test!(infile_comments, "infile_comments.lp");
generate_test!(infile_comments2, "infile_comments2.lp");
generate_test!(missing_signs, "missing_signs.lp");
generate_test!(scientific_notation_2, "scientific_notation_2.lp");
generate_test!(output, "output.lp");
generate_test!(output2_1, "output2_1.lp");
generate_test!(output2_2, "output2_2.lp");
generate_test!(output2_3, "output2_3.lp");
generate_test!(output2_4, "output2_4.lp");
generate_test!(complex_names, "complex_names.lp");

// Test files from various open source projects on Github

// Mosek: <https://docs.mosek.com/latest/capi/lp-format.html>
generate_test!(scientific_notation, "scientific_notation.lp");
generate_test!(mosek, "mosek.lp");
generate_test!(mosek_bounds, "mosek_bounds.lp");
generate_test!(lol, "lol.lp");
generate_test!(milo1, "milo1.lp");

// From <https://github.com/asbestian/jplex>
generate_test!(obj3_2cons, "3obj_2cons.lp");

// generate_test!(no_end_section, "no_end_section.lp");
generate_test!(obj_2cons_only_binary_vars, "2obj_2cons_only_binary_vars.lp");
generate_test!(obj_2cons_all_variable_types, "2obj_2cons_all_variable_types.lp");
generate_test!(obj_1cons_all_variables_with_bounds, "1obj_1cons_all_variables_with_bounds.lp");
generate_test!(afiro, "afiro.lp");
generate_test!(afiro_ext, "afiro_ext.lp");
generate_test!(boeing1, "boeing1.lp");
generate_test!(boeing2, "boeing2.lp");
generate_test!(fit1d, "fit1d.lp");
generate_test!(kb2, "kb2.lp");
generate_test!(sc50a, "sc50a.lp");

// From <https://github.com/odow/LPWriter.jl>
generate_test!(model2, "model2.lp");
#[test]
fn corrupt() {
    let input = read_file_from_resources("corrupt.lp").expect("failed to read file from resources");
    assert!(LpProblem::parse(&input).is_err());
}

// From <https://github.com/brymck/lp-parse/tree>
generate_test!(model, "model.lp");

// From <https://github.com/josephcslater/JupyterExamples>
generate_test!(sudoku, "sudoku.lp");

// From <https://github.com/afaf-taik/vehicularFL>
generate_test!(wbm, "wbm.lp");

// From <https://github.com/IBMDecisionOptimization/cplexrunonwml>
generate_test!(diet, "diet.lp");

// From <https://github.com/claudiosa/CCS>
// generate_test!(output_cplex_2, "output_cplex_2.lp");

// From <https://github.com/coin-or/pulp>
generate_test!(american_steel_problem, "AmericanSteelProblem.lp");
generate_test!(beer_distribution_problem, "BeerDistributionProblem.lp");
generate_test!(computer_plant_problem, "ComputerPlantProblem.lp");
generate_test!(whiskas_model_2, "WhiskasModel2.lp");

// From <https://lpsolve.sourceforge.net/5.0/CPLEX-format.htm>
generate_test!(cplex, "cplex.lp");
