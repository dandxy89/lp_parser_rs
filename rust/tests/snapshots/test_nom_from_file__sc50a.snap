---
source: rust/tests/test_nom_from_file.rs
expression: parsed
---
name: "Problem name: sc50a.lp"
sense: Minimize
objectives:
  MAXIM:
    name: MAXIM
    coefficients:
      - name: COL00004
        value: -1
constraints:
  ROW00001:
    type: Standard
    name: ROW00001
    coefficients:
      - name: COL00001
        value: 2
      - name: COL00002
        value: 1
      - name: COL00003
        value: 1.5
    operator: LTE
    rhs: 170
  ROW00002:
    type: Standard
    name: ROW00002
    coefficients:
      - name: COL00001
        value: 1
      - name: COL00002
        value: 2
      - name: COL00003
        value: 1.5
    operator: LTE
    rhs: 130
  ROW00004:
    type: Standard
    name: ROW00004
    coefficients:
      - name: COL00004
        value: 1
      - name: COL00005
        value: -1
    operator: EQ
    rhs: 0
  ROW00005:
    type: Standard
    name: ROW00005
    coefficients:
      - name: COL00001
        value: -1
      - name: COL00006
        value: 1
    operator: EQ
    rhs: 0
  ROW00006:
    type: Standard
    name: ROW00006
    coefficients:
      - name: COL00002
        value: -1
      - name: COL00007
        value: 1
    operator: EQ
    rhs: 0
  ROW00007:
    type: Standard
    name: ROW00007
    coefficients:
      - name: COL00003
        value: -1
      - name: COL00008
        value: 1
    operator: EQ
    rhs: 0
  ROW00008:
    type: Standard
    name: ROW00008
    coefficients:
      - name: COL00006
        value: -1
      - name: COL00009
        value: 1
    operator: LTE
    rhs: 0
  ROW00009:
    type: Standard
    name: ROW00009
    coefficients:
      - name: COL00007
        value: -1
      - name: COL00010
        value: 1
    operator: LTE
    rhs: 0
  ROW00010:
    type: Standard
    name: ROW00010
    coefficients:
      - name: COL00008
        value: -1
      - name: COL00011
        value: 1
    operator: LTE
    rhs: 0
  ROW00011:
    type: Standard
    name: ROW00011
    coefficients:
      - name: COL00009
        value: -1
      - name: COL00012
        value: 2
      - name: COL00013
        value: 1
      - name: COL00014
        value: 1.5
    operator: LTE
    rhs: 170
  ROW00012:
    type: Standard
    name: ROW00012
    coefficients:
      - name: COL00010
        value: -1
      - name: COL00012
        value: 1
      - name: COL00013
        value: 2
      - name: COL00014
        value: 1.5
    operator: LTE
    rhs: 130
  ROW00013:
    type: Standard
    name: ROW00013
    coefficients:
      - name: COL00005
        value: 1
      - name: COL00011
        value: -1
    operator: LTE
    rhs: 0
  ROW00014:
    type: Standard
    name: ROW00014
    coefficients:
      - name: COL00004
        value: 1.1
      - name: COL00015
        value: -1
    operator: EQ
    rhs: 0
  ROW00015:
    type: Standard
    name: ROW00015
    coefficients:
      - name: COL00005
        value: 1
      - name: COL00015
        value: 1
      - name: COL00016
        value: -1
    operator: EQ
    rhs: 0
  ROW00016:
    type: Standard
    name: ROW00016
    coefficients:
      - name: COL00006
        value: -1
      - name: COL00012
        value: -1
      - name: COL00017
        value: 1
    operator: EQ
    rhs: 0
  ROW00017:
    type: Standard
    name: ROW00017
    coefficients:
      - name: COL00007
        value: -1
      - name: COL00013
        value: -1
      - name: COL00018
        value: 1
    operator: EQ
    rhs: 0
  ROW00018:
    type: Standard
    name: ROW00018
    coefficients:
      - name: COL00008
        value: -1
      - name: COL00014
        value: -1
      - name: COL00019
        value: 1
    operator: EQ
    rhs: 0
  ROW00019:
    type: Standard
    name: ROW00019
    coefficients:
      - name: COL00017
        value: -1
      - name: COL00020
        value: 1
    operator: LTE
    rhs: 0
  ROW00020:
    type: Standard
    name: ROW00020
    coefficients:
      - name: COL00018
        value: -1
      - name: COL00021
        value: 1
    operator: LTE
    rhs: 0
  ROW00021:
    type: Standard
    name: ROW00021
    coefficients:
      - name: COL00019
        value: -1
      - name: COL00022
        value: 1
    operator: LTE
    rhs: 0
  ROW00022:
    type: Standard
    name: ROW00022
    coefficients:
      - name: COL00020
        value: -1
      - name: COL00023
        value: 2
      - name: COL00024
        value: 1
      - name: COL00025
        value: 1.5
    operator: LTE
    rhs: 170
  ROW00023:
    type: Standard
    name: ROW00023
    coefficients:
      - name: COL00021
        value: -1
      - name: COL00023
        value: 1
      - name: COL00024
        value: 2
      - name: COL00025
        value: 1.5
    operator: LTE
    rhs: 130
  ROW00024:
    type: Standard
    name: ROW00024
    coefficients:
      - name: COL00016
        value: 1
      - name: COL00022
        value: -1
    operator: LTE
    rhs: 0
  ROW00025:
    type: Standard
    name: ROW00025
    coefficients:
      - name: COL00015
        value: 1.1
      - name: COL00026
        value: -1
    operator: EQ
    rhs: 0
  ROW00026:
    type: Standard
    name: ROW00026
    coefficients:
      - name: COL00016
        value: 1
      - name: COL00026
        value: 1
      - name: COL00027
        value: -1
    operator: EQ
    rhs: 0
  ROW00027:
    type: Standard
    name: ROW00027
    coefficients:
      - name: COL00017
        value: -1
      - name: COL00023
        value: -1
      - name: COL00028
        value: 1
    operator: EQ
    rhs: 0
  ROW00028:
    type: Standard
    name: ROW00028
    coefficients:
      - name: COL00018
        value: -1
      - name: COL00024
        value: -1
      - name: COL00029
        value: 1
    operator: EQ
    rhs: 0
  ROW00029:
    type: Standard
    name: ROW00029
    coefficients:
      - name: COL00019
        value: -1
      - name: COL00025
        value: -1
      - name: COL00030
        value: 1
    operator: EQ
    rhs: 0
  ROW00030:
    type: Standard
    name: ROW00030
    coefficients:
      - name: COL00028
        value: -1
      - name: COL00031
        value: 1
    operator: LTE
    rhs: 0
  ROW00031:
    type: Standard
    name: ROW00031
    coefficients:
      - name: COL00029
        value: -1
      - name: COL00032
        value: 1
    operator: LTE
    rhs: 0
  ROW00032:
    type: Standard
    name: ROW00032
    coefficients:
      - name: COL00030
        value: -1
      - name: COL00033
        value: 1
    operator: LTE
    rhs: 0
  ROW00033:
    type: Standard
    name: ROW00033
    coefficients:
      - name: COL00031
        value: -1
      - name: COL00034
        value: 2
      - name: COL00035
        value: 1
      - name: COL00036
        value: 1.5
    operator: LTE
    rhs: 170
  ROW00034:
    type: Standard
    name: ROW00034
    coefficients:
      - name: COL00032
        value: -1
      - name: COL00034
        value: 1
      - name: COL00035
        value: 2
      - name: COL00036
        value: 1.5
    operator: LTE
    rhs: 130
  ROW00035:
    type: Standard
    name: ROW00035
    coefficients:
      - name: COL00027
        value: 1
      - name: COL00033
        value: -1
    operator: LTE
    rhs: 0
  ROW00036:
    type: Standard
    name: ROW00036
    coefficients:
      - name: COL00026
        value: 1.1
      - name: COL00037
        value: -1
    operator: EQ
    rhs: 0
  ROW00037:
    type: Standard
    name: ROW00037
    coefficients:
      - name: COL00027
        value: 1
      - name: COL00037
        value: 1
      - name: COL00038
        value: -1
    operator: EQ
    rhs: 0
  ROW00038:
    type: Standard
    name: ROW00038
    coefficients:
      - name: COL00028
        value: -1
      - name: COL00034
        value: -1
      - name: COL00039
        value: 1
    operator: EQ
    rhs: 0
  ROW00039:
    type: Standard
    name: ROW00039
    coefficients:
      - name: COL00029
        value: -1
      - name: COL00035
        value: -1
      - name: COL00040
        value: 1
    operator: EQ
    rhs: 0
  ROW00040:
    type: Standard
    name: ROW00040
    coefficients:
      - name: COL00030
        value: -1
      - name: COL00036
        value: -1
      - name: COL00041
        value: 1
    operator: EQ
    rhs: 0
  ROW00041:
    type: Standard
    name: ROW00041
    coefficients:
      - name: COL00039
        value: -1
      - name: COL00042
        value: 1
    operator: LTE
    rhs: 0
  ROW00042:
    type: Standard
    name: ROW00042
    coefficients:
      - name: COL00040
        value: -1
      - name: COL00043
        value: 1
    operator: LTE
    rhs: 0
  ROW00043:
    type: Standard
    name: ROW00043
    coefficients:
      - name: COL00041
        value: -1
      - name: COL00044
        value: 1
    operator: LTE
    rhs: 0
  ROW00044:
    type: Standard
    name: ROW00044
    coefficients:
      - name: COL00042
        value: -1
      - name: COL00045
        value: 2
      - name: COL00046
        value: 1
      - name: COL00047
        value: 1.5
    operator: LTE
    rhs: 170
  ROW00045:
    type: Standard
    name: ROW00045
    coefficients:
      - name: COL00043
        value: -1
      - name: COL00045
        value: 1
      - name: COL00046
        value: 2
      - name: COL00047
        value: 1.5
    operator: LTE
    rhs: 130
  ROW00046:
    type: Standard
    name: ROW00046
    coefficients:
      - name: COL00038
        value: 1
      - name: COL00044
        value: -1
    operator: LTE
    rhs: 0
  ROW00047:
    type: Standard
    name: ROW00047
    coefficients:
      - name: COL00037
        value: 1.1
      - name: COL00048
        value: -1
    operator: EQ
    rhs: 0
  ROW00048:
    type: Standard
    name: ROW00048
    coefficients:
      - name: COL00045
        value: -0.8
      - name: COL00046
        value: 0.1
      - name: COL00047
        value: 0.15
    operator: LTE
    rhs: 0
  ROW00049:
    type: Standard
    name: ROW00049
    coefficients:
      - name: COL00045
        value: 0.1
      - name: COL00046
        value: -0.8
      - name: COL00047
        value: 0.15
    operator: LTE
    rhs: 0
  ROW00050:
    type: Standard
    name: ROW00050
    coefficients:
      - name: COL00047
        value: -1
      - name: COL00048
        value: 1
    operator: LTE
    rhs: 0
variables:
  COL00001:
    name: COL00001
    var_type: Free
  COL00002:
    name: COL00002
    var_type: Free
  COL00003:
    name: COL00003
    var_type: Free
  COL00004:
    name: COL00004
    var_type: Free
  COL00005:
    name: COL00005
    var_type: Free
  COL00006:
    name: COL00006
    var_type: Free
  COL00007:
    name: COL00007
    var_type: Free
  COL00008:
    name: COL00008
    var_type: Free
  COL00009:
    name: COL00009
    var_type: Free
  COL00010:
    name: COL00010
    var_type: Free
  COL00011:
    name: COL00011
    var_type: Free
  COL00012:
    name: COL00012
    var_type: Free
  COL00013:
    name: COL00013
    var_type: Free
  COL00014:
    name: COL00014
    var_type: Free
  COL00015:
    name: COL00015
    var_type: Free
  COL00016:
    name: COL00016
    var_type: Free
  COL00017:
    name: COL00017
    var_type: Free
  COL00018:
    name: COL00018
    var_type: Free
  COL00019:
    name: COL00019
    var_type: Free
  COL00020:
    name: COL00020
    var_type: Free
  COL00021:
    name: COL00021
    var_type: Free
  COL00022:
    name: COL00022
    var_type: Free
  COL00023:
    name: COL00023
    var_type: Free
  COL00024:
    name: COL00024
    var_type: Free
  COL00025:
    name: COL00025
    var_type: Free
  COL00026:
    name: COL00026
    var_type: Free
  COL00027:
    name: COL00027
    var_type: Free
  COL00028:
    name: COL00028
    var_type: Free
  COL00029:
    name: COL00029
    var_type: Free
  COL00030:
    name: COL00030
    var_type: Free
  COL00031:
    name: COL00031
    var_type: Free
  COL00032:
    name: COL00032
    var_type: Free
  COL00033:
    name: COL00033
    var_type: Free
  COL00034:
    name: COL00034
    var_type: Free
  COL00035:
    name: COL00035
    var_type: Free
  COL00036:
    name: COL00036
    var_type: Free
  COL00037:
    name: COL00037
    var_type: Free
  COL00038:
    name: COL00038
    var_type: Free
  COL00039:
    name: COL00039
    var_type: Free
  COL00040:
    name: COL00040
    var_type: Free
  COL00041:
    name: COL00041
    var_type: Free
  COL00042:
    name: COL00042
    var_type: Free
  COL00043:
    name: COL00043
    var_type: Free
  COL00044:
    name: COL00044
    var_type: Free
  COL00045:
    name: COL00045
    var_type: Free
  COL00046:
    name: COL00046
    var_type: Free
  COL00047:
    name: COL00047
    var_type: Free
  COL00048:
    name: COL00048
    var_type: Free
